### PR TITLE
python: reference count python-option

### LINKED ISSUE #4566: Multiple usage of a python parser causes segmentation fault on shutdown
# syslog-ng
## Version of syslog-ng
4.3.0

# Issue
If a python-parser is used more than once, syslog-ng can cnot shut down normally and coredumps on shutdown.

## Backtrace(for the `syslog-ng-ctl stop` case):
```
Thread 1 "syslog-ng" received signal SIGSEGV, Segmentation fault.
0x00007ffff72c7c5e in free () from /usr/lib/libc.so.6
(gdb) bt full
#0  0x00007ffff72c7c5e in free () from /usr/lib/libc.so.6
No symbol table info available.
#1  0x00007ffff6f3198c in python_option_free (s=0x55555568e380) at ../modules/python/python-options.c:82
No locals.
#2  0x00007ffff7509b00 in g_list_foreach () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#3  0x00007ffff7509ef0 in g_list_free_full () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#4  0x00007ffff6f31fa9 in python_options_free (self=0x55555570afa0) at ../modules/python/python-options.c:348
No locals.
#5  0x00007ffff6f2a190 in python_parser_free (d=0x5555556d7370) at ../modules/python/python-logparser.c:297
        self = 0x5555556d7370
        gstate = PyGILState_UNLOCKED
#6  0x00007ffff7ea9895 in _free (self=0x5555556d7370) at ../lib/logpipe.c:117
No locals.
#7  0x00007ffff7ea9961 in log_pipe_unref (self=0x5555556d7370) at ../lib/logpipe.c:132
        __func__ = "log_pipe_unref"
#8  0x00007ffff7e98e1b in _log_expr_node_free (self=0x555555706830) at ../lib/cfg-tree.c:306
        next = 0x555555706830
        p = 0x0
#9  0x00007ffff7e9889d in log_expr_node_unref (self=0x555555706830) at ../lib/cfg-tree.c:53
        __func__ = "log_expr_node_unref"
#10 0x00007ffff7e98ddd in _log_expr_node_free (self=0x55555570b0b0) at ../lib/cfg-tree.c:303
        next = 0x0
        p = 0x555555706830
#11 0x00007ffff7e9889d in log_expr_node_unref (self=0x55555570b0b0) at ../lib/cfg-tree.c:53
--Type <RET> for more, q to quit, c to continue without paging--
        __func__ = "log_expr_node_unref"
#12 0x00007ffff74f8c23 in ?? () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#13 0x00007ffff74f98b0 in g_hash_table_remove_all () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#14 0x00007ffff74fce73 in g_hash_table_destroy () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#15 0x00007ffff7e9b8c4 in cfg_tree_free_instance (self=0x5555555a3858) at ../lib/cfg-tree.c:1679
No locals.
#16 0x00007ffff7e91ec1 in cfg_free (self=0x5555555a3680) at ../lib/cfg.c:738
        __func__ = "cfg_free"
#17 0x00007ffff7eb9caf in main_loop_free_config (self=0x7ffff7fbdc80 <main_loop>) at ../lib/mainloop.c:732
No locals.
#18 0x00007ffff7eb9cd9 in main_loop_deinit (self=0x7ffff7fbdc80 <main_loop>) at ../lib/mainloop.c:739
No locals.
#19 0x0000555555556b08 in main (argc=1, argv=0x7fffffffe3d8) at ../syslog-ng/main.c:338
        rc = 0
        ctx = 0x555555560880
        error = 0x0
        main_loop = 0x7ffff7fbdc80 <main_loop>
        exit_before_main_loop_run = 0
```

## Steps to reproduce

1. Start syslog-ng `syslog-ng -Fdev` with the config bellow
2. Stop syslog-ng, for example with `syslog-ng-ctl stop`.

After these step syslog-ng will dump it's core:
```
[2023-07-27T09:38:07.947800] syslog-ng shutting down; version='4.3.0'
[2023-07-27T09:38:07.947882] EOF on control channel, closing connection;
Segmentation fault: oops, process './install/sbin/syslog-ng' core dumped
```

## Configuration
```
@version: 4.3
options {
    stats_level(1);
};
source s_file_1abd67defa7b43d98cb30f5fbb939974 {
    file("/tmp/a" flags(no-parse));
};
parser p_python_parser_fa6b0e03632345f6a27896a4ac53096c {
    python(class(MyParser)
 options("info", "This is MyParser"));
};
destination d_file_f74e821008cd4c9684114d5ffd74728e {
    file("/tmp/b");
};
log {
source(s_file_1abd67defa7b43d98cb30f5fbb939974);
parser(p_python_parser_fa6b0e03632345f6a27896a4ac53096c);
parser(p_python_parser_fa6b0e03632345f6a27896a4ac53096c);
destination(d_file_f74e821008cd4c9684114d5ffd74728e);

flags(flow-control);
};

python {

from syslogng import LogParser, Logger

class MyParser(LogParser):
    logger = Logger()

    def parse(self, log_message):
        self.logger.error("Class MyParser parse, msg='{}'".format(log_message.get_as_str("MESSAGE")))
        return True
};
```

### PR DESCRIPTION
**Pull request by alltilla**
`python_options_add_options()` took the ownership of the `PythonOption`, resulting in double frees in case of cloning.

Fixes #4566

### CODE REVIEW

File: modules/python/python-options.c [hunk 1]
```diff
@@ -73,14 +77,31 @@ python_option_create_value_py_object(const PythonOption *s)
...
+python_option_ref(PythonOption *s)
+{
+  g_assert(!s || g_atomic_counter_get(&s->ref_cnt) > 0);
+
+  if (s)
+    {
+      g_atomic_counter_inc(&s->ref_cnt);
+    }
+  return s;
+}
```
Comment (MrAnno) [inline modules/python/python-options.c:90]: Why do need reference counting for PythonOption.
The ownership of this should be something simple in my opinion.
A PythonOption should be owned by its container class (PythonOptions), which should belong to a driver.

Accumulating multiple options() block can also be done without creating multiple Options instances, as far as I see (`python_options_add_options()` sounds weird)

Comment (alltilla) [reply modules/python/python-options.c:90]: The problem is that a `PythonOptions` does not necessary belong to one driver. When we clone a parser we need to either clone `PythonOptions` with all its `PythonOption`, or make `PythonOptions` and `PythonOption` refcounted, or clone `PythonOptions` and make `PythonOption` refcounted. I did the latter.

Comment (alltilla) [reply modules/python/python-options.c:90]: Cloning `PythonOption` is overkill, as we need to write the clone fn for all types. We can fortunately refcount it as it does not change after constructing it.

**Review by MrAnno — DISMISSED** [review]
lgtm

Nevertheless, `python_options_add_options()` and the way it's used are confusing in my opinion.
We should accumulate a single instance instead of merging multiple of them.

**Comment by MrAnno** [comment]
A NEWS file update will be needed for the last 2 bugfixes.

**Comment by alltilla** [comment]
We should clear up all the confusion IMO, it caused enough problems. I can modify the grammar to accumulate all options from the options block to one options object. We need something for the parser clone fn still. I can add an options::clone fn which creates a new options and increases the refcount for each option inside. This makes it possible to add new option to the cloned options while saving memory and code complexity. What do you think?

**Comment by MrAnno** [comment]
Sounds good to me.

**Review by MrAnno — APPROVED** [review]
Thank you very much! :)

**Pull request merged**